src/main/java/org/commcare/xml/EndpointParser.java

```diff
@@ -31,16 +35,23 @@ public Endpoint parse() throws InvalidStructureException, IOException, XmlPullPa
         }
 
         Vector<StackOperation> stackOperations = new Vector<>();
-        Vector<String> arguments = new Vector<>();
+        Vector<EndpointArgument> arguments = new Vector<>();
 
         while (nextTagInBlock(NAME_ENDPOINT)) {
             String tagName = parser.getName().toLowerCase();
             if (tagName.contentEquals(NAME_ARGUMENT)) {
-                String argumentID = parser.getAttributeValue(null, "id");
+                String argumentID = parser.getAttributeValue(null, ATTR_ARGUMENT_ID);
                 if (argumentID == null || argumentID.isEmpty()) {
                     throw new InvalidStructureException("argument must define a non empty id", parser);
                 }
-                arguments.add(argumentID);
+                String argInstanceId = parser.getAttributeValue(null, ATTR_ARGUMENT_INSTANCE_ID);
+                String argInstanceSrc = parser.getAttributeValue(null, ATTR_ARGUMENT_INSTANCE_SRC);
+
+                if (argInstanceId != null && argInstanceSrc == null) {
+                    throw new InvalidStructureException(
+                            "argument containing a non-null instance-id must define an instance-src", parser);
+                }
+                arguments.add(new EndpointArgument(argumentID, argInstanceId, argInstanceSrc));
```

Comment:
Just to make sure I understand correctly, is the idea here that `EndpointArgument` also handles single case selection where `<argument id="case_id"/>`? So if there is no `argInstanceId ` and no `argInstanceSrc ` the functionality from this point is essentially the same as before.

Reply:
yes, spot on.